src/objc/AudioPreprocessing.m

```diff
@@ -0,0 +1,370 @@
+    }
+}
+
+@end // TCSoundClassifierPowerSpectrum
+
+
+@interface TCSoundClassifierMelFrequencyFilterBank : NSObject
+
+-(nullable instancetype)initWithMinFrequency:(float)minFreq
+                                maxFrequency:(float)maxFrequency
+                                  sampleRate:(double)sampleRate
+                                 numMelBands:(NSUInteger)numMelBands
+                                     numBins:(NSUInteger)numBins;
+
+-(void)apply:(const float*)powerSpectrum
+      output:(float *)out;
+
+@end
+
+
+@implementation TCSoundClassifierMelFrequencyFilterBank {
+    float _minFreq, _maxFreq, _sampleRate;
+    NSUInteger _numMelBands, _numBins;
+    NSMutableArray* _filters;
+    float** _filterBanks;
+}
+
+-(nullable instancetype)initWithMinFrequency:(float)minFreq
+                                maxFrequency:(float)maxFreq
+                                  sampleRate:(double)sampleRate
+                                 numMelBands:(NSUInteger)numMelBands
+                                     numBins:(NSUInteger)numBins {
+
+    self = [super init];
+    if (!self) return nil;
+
+    _minFreq = minFreq;
+    _maxFreq = maxFreq;
+    _sampleRate = sampleRate;
+    _numMelBands = numMelBands;
+    _numBins = numBins;
+    _filterBanks = calloc(_numMelBands, sizeof(float *));
+
+    NSAssert(_minFreq < _maxFreq, @"Min mel frequency must be less than max mel frequency.");
+    NSAssert(_maxFreq != 0, @"Max frequency cannot be 0.");
+    NSAssert(_sampleRate != 0, @"Sample rate cannot be 0.");
+    NSAssert(_numMelBands != 0, @"numMelBands cannot be 0.");
+    NSAssert(_numBins != 0, @"numBins cannot be 0.");
+
+    _filters = [[NSMutableArray alloc] initWithCapacity:numMelBands];
+
+    float melMin = hzToMel(_minFreq);
+    float melMax = hzToMel(_maxFreq);
+
+    // Get points equally spaced in the Mel Scale.
+    float melBandEdges[_numMelBands + 2];
+    float increment = (melMax - melMin) / (_numMelBands + 1);
+    vDSP_vramp(&melMin, &increment, melBandEdges, 1, _numMelBands + 2);
+
+    // Get points equally spaced in the Hz Scale.
+    float spectrogramBins[spectrumSize];
+    float minHz = 0.0;
+    float HzIncrement = (sampleRate/2) / (spectrumSize - 1);
+    vDSP_vramp(&minHz, &HzIncrement, spectrogramBins, 1, spectrumSize);
+
+    // Convert Hz to Mel
+    for(int i = 0; i < spectrumSize; i++) {
+        spectrogramBins[i] = hzToMel(spectrogramBins[i]);
+    }
+
+    // For each band edge, create a new filter (basically a float array that we vDSP_dot later)
+    for(size_t i = 0; i < _numMelBands; i++) {
+        // Grab the low, center, and bottom edges
+        float lower_edge = melBandEdges[i];
+        float center_edge = melBandEdges[i+1];
+        float upper_edge = melBandEdges[i+2];
+
+        // Compute lower and upper slopes
+        float* lowerBank = (float*)calloc(spectrumSize, sizeof(float));
+        float* upperBank = (float*)calloc(spectrumSize, sizeof(float));
```

Comment:
I don't think lowerBank and upperBank are being freed anywhere, right?

Reply:
Yes, you're right. Fixed. Thanks.